RecoMuon/MuonIdentification/plugins/cuts/MuonPOGStandardCuts.cc

```diff
@@ -0,0 +1,69 @@
+#include "PhysicsTools/SelectorUtils/interface/CutApplicatorBase.h"
+#include "PhysicsTools/SelectorUtils/interface/CutApplicatorWithEventContentBase.h"
+#include "DataFormats/MuonReco/interface/Muon.h"
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+
+class MuonPOGStandardCut : public CutApplicatorWithEventContentBase
+{
+public:
+  MuonPOGStandardCut(const edm::ParameterSet& c);
+
+  result_type operator()(const reco::MuonPtr&) const override final;
+  CandidateType candidateType() const override final { return MUON; }
+
+  void setConsumes(edm::ConsumesCollector&) override final;
+  void getEventContent(const edm::EventBase&) override final;
+
+private:
+  enum CutType {
+    LOOSE, MEDIUM, TIGHT, SOFT, HIGHPT,
+    NONE
+  } cutType_;
+  edm::Handle<reco::VertexCollection> vtxs_;
+};
+DEFINE_EDM_PLUGIN(CutApplicatorFactory,
+                  MuonPOGStandardCut, "MuonPOGStandardCut");
+
+// Define constructors and initialization routines
+MuonPOGStandardCut::MuonPOGStandardCut(const edm::ParameterSet& c):
+  CutApplicatorWithEventContentBase(c)
+{
+  const auto cutTypeName = c.getParameter<std::string>("idName");
+  if ( cutTypeName == "loose") cutType_ = LOOSE;
+  else if ( cutTypeName == "tight") cutType_ = TIGHT;
+  else if ( cutTypeName == "medium") cutType_ = MEDIUM;
+  else if ( cutTypeName == "soft" ) cutType_ = SOFT;
+  else if ( cutTypeName == "highpt" ) cutType_ = HIGHPT;
+  else
+  {
+    edm::LogError("MuonPOGStandardCut") << "Wrong cut id name, " << cutTypeName;
+    cutType_ = NONE;
+  }
+
+  contentTags_.emplace("vertices", c.getParameter<edm::InputTag>("vertexSrc"));
+}
+
+void MuonPOGStandardCut::setConsumes(edm::ConsumesCollector& cc)
+{
+  auto vtcs = cc.consumes<reco::VertexCollection>(contentTags_["vertices"]);
+  contentTokens_.emplace("vertices", vtcs);
+}
+
+void MuonPOGStandardCut::getEventContent(const edm::EventBase& ev)
+{
+  ev.getByLabel(contentTags_["vertices"], vtxs_);
+}
+
+// Functors for evaluation
+CutApplicatorBase::result_type MuonPOGStandardCut::operator()(const reco::MuonPtr& cand) const
+{
+  if ( cutType_ == LOOSE ) return muon::isLooseMuon(*cand);
+  else if ( cutType_ == TIGHT ) return muon::isTightMuon(*cand, vtxs_->at(0));
```

Comment:
Hi Carl,

The code is correct. edm::Handle will throw an exception if this occurs,
not crashing and generating a stack trace..
This what we wanted to happen in this case, since it indicates a poorly
defined dataformat.

Is there a serious problem aside from style?

-Lindsey

On Mon, Jun 1, 2015 at 6:50 PM, Carl Vuosalo <notifications@github.com>
wrote:

> In RecoMuon/MuonIdentification/plugins/cuts/MuonPOGStandardCuts.cc
> <https://github.com/cms-sw/cmssw/pull/9273#discussion_r31444916>:
>
> > +void MuonPOGStandardCut::setConsumes(edm::ConsumesCollector& cc)
> > +{
> > +  auto vtcs = cc.consumes<reco::VertexCollection>(contentTags_["vertices"]);
> > +  contentTokens_.emplace("vertices", vtcs);
> > +}
> > +
> > +void MuonPOGStandardCut::getEventContent(const edm::EventBase& ev)
> > +{
> > +  ev.getByLabel(contentTags_["vertices"], vtxs_);
> > +}
> > +
> > +// Functors for evaluation
> > +CutApplicatorBase::result_type MuonPOGStandardCut::operator()(const reco::MuonPtr& cand) const
> > +{
> > +  if ( cutType_ == LOOSE ) return muon::isLooseMuon(*cand);
> > +  else if ( cutType_ == TIGHT ) return muon::isTightMuon(*cand, vtxs_->at(0));
>
> For safety, please check vtxs_.isValid() before dereferencing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9273/files#r31444916>.
>